buildSrc/src/main/kotlin/com/datadog/gradle/plugin/gitclone/GitCloneDependenciesTask.kt

```diff
@@ -0,0 +1,159 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-2020 Datadog, Inc.
+ */
+
+package com.datadog.gradle.plugin.gitclone
+
+import com.datadog.gradle.utils.execShell
+import java.io.File
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+
+open class GitCloneDependenciesTask : DefaultTask() {
+
+    internal var extension: GitCloneDependenciesExtension =
+        GitCloneDependenciesExtension()
+
+    init {
+        group = "datadog"
+        description = "Clones the source code of a dependency into this module"
+        outputs.upToDateWhen { false }
+    }
+
+    // region Task
+
+    @TaskAction
+    fun applyTask() {
+        extension.dependencies.forEach {
+            cloneDependency(it)
+        }
+    }
+
+    private fun cloneDependency(
+        dependency: GitCloneDependenciesExtension.Dependency
+    ) {
+        val target = createTempDir()
+        val copyFrom = if (dependency.originSubFolder.isEmpty()) {
+            target
+        } else {
+            File("${target.absolutePath}${File.separator}${dependency.originSubFolder}")
+        }
+        cloneRepository(dependency, target)
+        copySources(copyFrom, project.projectDir, dependency.excludedPrefixes)
+        deleteClone(target)
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun cloneRepository(
+        dependency: GitCloneDependenciesExtension.Dependency,
+        target: File
+    ) {
+        println(" --- Cloning ${dependency.originRepository} into ${target.absolutePath}")
+        project.execShell("git", "clone", "--depth", "1", dependency.originRepository, target.absolutePath)
+        println(" --- Cloned")
+    }
+
+    private fun copySources(
+        src: File,
+        dest: File,
+        excludedPrefixes: List<String>
+    ) {
+        val srcPath = "${src.absolutePath}${File.separator}"
+        val destPath = "${dest.absolutePath}${File.separator}"
+        println(" --- Copying from $srcPath into $destPath")
+
+        copyDirectory(src, srcPath, destPath, excludedPrefixes)
+
+        println(" --- Copied")
+    }
+
+    private fun copyDirectory(
+        dir: File,
+        srcPath: String,
+        destPath: String,
+        excludedPrefixes: List<String>
+    ) {
+        val children = dir.listFiles().orEmpty()
```

Comment:
you can use `dir.walkBottomUp()` which returns a sequence and avoid the recursion

Reply:
Oh nice one, didn't know that :)